demos/device_defender_for_aws/defender_demo.c

```diff
@@ -0,0 +1,649 @@
+#include "iot_config.h"
+
+/* Demo logging include. */
+#include "iot_demo_logging.h"
+
+/* Demo config. */
+#include "demo_config.h"
+
+/* Metrics collector. */
+#include "metrics_collector.h"
+
+/* Report builder. */
+#include "report_builder.h"
+
+/* MQTT operations. */
+#include "mqtt_demo_helpers.h"
+
+/* JSON Library. */
+#include "core_json.h"
+
+/* Device Defender Client Library. */
+#include "defender.h"
+
+
+#ifndef THING_NAME
+
+/**
+ * @brief Predefined thing name.
+ *
+ * This is the example predefine thing name and could be compiled in ROM code.
+ */
+    #define THING_NAME    clientcredentialIOT_THING_NAME
+#endif
+
+
+/**
+ * @brief The length of #THING_NAME.
+ */
+#define THING_NAME_LENGTH                      ( ( uint16_t ) ( sizeof( THING_NAME ) - 1 ) )
+
+
+/**
+ * @brief Number of seconds to wait for the response from AWS IoT Device
+ * Defender service.
+ */
+#define DEFENDER_RESPONSE_WAIT_SECONDS    ( 2 )
+
+
+/**
+ * @brief Status values of the device defender report.
+ */
+typedef enum
+{
+    ReportStatusNotReceived,
+    ReportStatusAccepted,
+    ReportStatusRejected
+} ReportStatus_t;
+/*-----------------------------------------------------------*/
+
+
+/**
+ * @brief The MQTT context used for MQTT operation.
+ */
+static MQTTContext_t xMqttContext;
+
+/**
+ * @brief The network context used for Openssl operation.
+ */
+static NetworkContext_t xNetworkContext;
+
+/**
+ * @brief Static buffer used to hold MQTT messages being sent and received.
+ */
+static uint8_t ucSharedBuffer[ NETWORK_BUFFER_SIZE ];
+
+/**
+ * @brief Static buffer used to hold MQTT messages being sent and received.
+ */
+static MQTTFixedBuffer_t xBuffer =
+{
+    .pBuffer = ucSharedBuffer,
+    .size    = NETWORK_BUFFER_SIZE
+};
+
+/**
+ * @brief Network Stats.
+ */
+static NetworkStats_t networkStats;
+
+/**
+ * @brief Open TCP ports array.
+ */
+static uint16_t openTcpPorts[ OPEN_TCP_PORTS_ARRAY_SIZE ];
+
+/**
+ * @brief Open UDP ports array.
+ */
+static uint16_t openUdpPorts[ OPEN_UDP_PORTS_ARRAY_SIZE ];
+
+/**
+ * @brief Established connections array.
+ */
+static Connection_t establishedConnections[ ESTABLISHED_CONNECTIONS_ARRAY_SIZE ];
+
+/**
+ * @brief All the metrics sent in the device defender report.
+ */
+static ReportMetrics_t deviceMetrics;
+
+/**
+ * @brief Report status.
+ */
+static ReportStatus_t reportStatus;
+
+/**
+ * @brief Buffer for generating the device defender report.
+ */
+static char deviceMetricsJsonReport[ DEVICE_METRICS_REPORT_BUFFER_SIZE ];
+
+/**
+ * @brief Report Id sent in the defender report.
+ */
+static uint32_t reportId = 0;
+/*-----------------------------------------------------------*/
+
+
+
+/**
+ * @brief Callback to receive the incoming publish messages from the MQTT broker.
+ *
+ * @param[in] pPublishInfo Pointer to publish info of the incoming publish.
+ * @param[in] packetIdentifier Packet identifier of the incoming publish.
+ */
+static void publishCallback( MQTTContext_t * pxMqttContext,
```

Comment:
This publish callback looks different than the one for the new library?

Reply:
https://github.com/aws/aws-iot-device-sdk-embedded-C/blob/main/demos/defender/defender_demo_json/defender_demo.c#L118

Reply:
Sorry I just noticed that the csdk is out of date and not you lol